workers/unity/Assets/Playground/Scripts/EntityInitialization/EntityGameObjectCreator.cs

```diff
@@ -1,53 +1,82 @@
-using System;
 using System.Collections.Generic;
+using System.IO;
+using Generated.Improbable;
 using Improbable.Gdk.Core;
-using Improbable.Worker;
-using Unity.Entities;
+using Improbable.Gdk.Core.GameObjectRepresentation;
 using UnityEngine;
-using Object = UnityEngine.Object;
+using Transform = Generated.Improbable.Transform.Transform;
 
 namespace Playground
 {
-    public class EntityGameObjectCreator
+    public class EntityGameObjectCreator : IEntityGameObjectCreator
     {
-        private readonly World world;
-        private readonly Dictionary<string, GameObject> cachedPrefabs;
+        private readonly Dictionary<string, GameObject> cachedPrefabs
+            = new Dictionary<string, GameObject>();
 
-        public EntityGameObjectCreator(World world, Dictionary<string, GameObject> initialCachedPrefabs = null)
+        private readonly string workerType;
+
+        public EntityGameObjectCreator(string workerType)
         {
-            this.world = world;
-            cachedPrefabs = initialCachedPrefabs ?? new Dictionary<string, GameObject>();
+            this.workerType = workerType;
         }
 
-        public GameObject CreateEntityGameObject(Entity entity, string prefabPath, Vector3 position,
-            Quaternion rotation, EntityId spatialEntityId)
+        public GameObject GetGameObjectForEntityAdded(SpatialOSEntity entity, WorkerSystem worker)
         {
-            if (!cachedPrefabs.TryGetValue(prefabPath, out var prefab))
+            if (!entity.HasComponent<Metadata.Component>())
+            {
+                return null;
+            }
+
+            var prefabName = entity.GetComponent<Metadata.Component>().EntityType;
+            if (!entity.HasComponent<Transform.Component>())
             {
-                prefab = Resources.Load<GameObject>(prefabPath);
+                cachedPrefabs[prefabName] = null;
+                return null;
+            }
+
+            var transform = entity.GetComponent<Transform.Component>();
+            var position = new Vector3(transform.Location.X, transform.Location.Y, transform.Location.Z) +
+                worker.Origin;
+            var rotation = new UnityEngine.Quaternion(transform.Rotation.X, transform.Rotation.Y,
+                transform.Rotation.Z, transform.Rotation.W);
+            var workerSpecificPath = Path.Combine("Prefabs", worker.WorkerType, prefabName);
+            var commonPath = Path.Combine("Prefabs", "Common", prefabName);
+
+            if (!cachedPrefabs.TryGetValue(workerSpecificPath, out var prefab)
```

Comment:
cache on prefabName instead of the path

Reply:
You could have multiple prefabs with the same name though in different paths. In fact I expect you would.

Reply:
This doesn't support that anyways. The prefab is either in the worker specific folder or the common folder.

Reply:
Ah true. This is per worker now (it wasn't before). In which case this can key on that.

Reply:
This is also a bug because I'm caching on name and then getting from path.... fixing